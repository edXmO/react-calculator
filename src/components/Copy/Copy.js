import React from 'react';
import { ReactComponent as GitLogo } from '../../assets/img/SVG/github-logo.svg';
const Copy = () => {
    return (
        <div className='copy'>
            <a className='copy__link' href='https://github.com/edXmO/react-calculator'>
                <GitLogo className='copy__icon' />
            </a>
            <p className='copy__text'>
                Built by edXmO, 2021
            </p>
        </div>
    )
}

export default Copy;