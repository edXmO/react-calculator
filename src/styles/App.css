.main-wrapper {
  height: 50rem;
  width: 30rem;
  border-radius: 10px;
  display: grid;
  margin-bottom: 3rem;
  grid-template-rows: 10% 25% auto auto auto;
  grid-template-columns: repeat(5, 1fr);
  grid-template-areas: "logo    logo    logo    logo    logo" "display display display display display" "buttons buttons buttons buttons buttons" "buttons buttons buttons buttons buttons" "buttons buttons buttons buttons buttons"; }

.logo {
  grid-area: logo;
  background-color: var(--color-background-btn);
  border-top-left-radius: 6px;
  border-top-right-radius: 6px; }

.screen {
  grid-area: display;
  background-color: var(--color-bg-white); }

.buttons {
  grid-area: buttons;
  background-color: var(--color-bg-btn);
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: repeat(5, 1fr); }

.copy {
  display: flex;
  flex-direction: column;
  align-items: center; }
  .copy__icon {
    width: 4.5rem;
    height: 4.5rem; }
  .copy__text {
    margin: 1rem 0;
    font-family: "Open Sans", sans-serif;
    font-style: italic;
    font-size: 1.7rem; }

.buttons {
  border-radius: 5px; }

:root {
  --color-bg-btn: #1c1c1c;
  --color-bg-white: #fff;
  --color-primary: #d4d4d2;
  --color-primary-btn-bg: #2b3333;
  --color-primary-btn-special: #b51442;
  --color-background-btn: #2d3749;
  --color-btns: #2d3749;
  --color-num-screen: #888c91;
  --color-bg-screen: #282e38;
  --color-btn-text: #dfe0e1; }

*,
*::before,
*::after {
  margin: 0;
  padding: 0;
  box-sizing: inherit; }

html {
  font-size: 62.5%; }

body {
  font-family: "Roboto", sans-serif;
  font-size: 1.6rem;
  line-height: 1.6rem;
  font-weight: 300;
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
  background: #396afc;
  /* fallback for old browsers */
  /* W3C, IE 10+/ Edge, Firefox 16+, Chrome 26+, Opera 12+, Safari 7+ */ }

.btn {
  align-self: center;
  justify-self: center;
  width: 100%;
  height: 100%;
  border: none;
  outline: none;
  font-size: 2rem;
  cursor: pointer;
  transition: all 0.3s; }
  .btn:nth-child(17) {
    border-bottom-left-radius: 5px; }
  .btn:nth-child(20) {
    border-bottom-right-radius: 5px; }
  .btn--operator {
    color: var(--color-btn-text);
    background-color: var(--color-primary-btn-special);
    height: 100%;
    width: 100%;
    border-radius: 0;
    box-shadow: inset 0 -2px 2px rgba(128, 16, 47, 0.829), inset -2px 0 2px rgba(128, 16, 47, 0.833); }
    .btn--operator:hover {
      color: var(--color-bg-white);
      background-color: #961136; }
      .btn--operator:hover:active {
        background-color: #b51442;
        box-shadow: inset -1px -3px 8px rgba(128, 16, 47, 0.829), inset -3px -1px 8px rgba(128, 16, 47, 0.833); }
  .btn--number {
    color: var(--color-btn-text);
    background-color: var(--color-background-btn);
    border-bottom: 1px solid rgba(255, 255, 255, 0.03);
    border-right: 1px solid rgba(255, 255, 255, 0.04);
    border-top: 0;
    border-left: 0; }
    .btn--number:hover {
      color: var(--color-bg-white);
      background-color: rgba(36, 44, 58, 0.8); }
      .btn--number:hover:active {
        background-color: rgba(45, 55, 73, 0.9);
        box-shadow: inset -1px -3px 5px rgba(36, 44, 58, 0.514), inset -3px -1px 5px rgba(36, 44, 58, 0.514); }

.screen {
  font-size: 4.5rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: var(--color-bg-screen); }
  .screen__currOperation {
    margin: 2rem 1rem;
    font-size: 3rem;
    color: var(--color-primary);
    align-self: flex-start; }
  .screen__display {
    color: var(--color-bg-white);
    margin: 2rem 1rem;
    align-self: flex-end; }
