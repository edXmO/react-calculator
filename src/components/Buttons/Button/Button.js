import React from 'react';
// Styles
import '../../../styles/App.css';
// redux
import { useDispatch, useSelector } from 'react-redux';
// Actions
import { clear_screen, del, lastKey_pressed, decimal, equals, addOperator } from '../../../actions';

const Button = ({ type, btn, operatorStyle }) => {

    const lastResult = useSelector(({ calculator }) => calculator.result);
    const currDisplay = useSelector(({ calculator }) => calculator.screen);
    const lastType = useSelector(({ calculator }) => calculator.lastType);
    const keyHistory = useSelector(({ calculator }) => calculator.keyHistory);
    const dispatch = useDispatch();

    const handleBtnClick = (btn, type, lastType) => {
        if (type === 'Display') {
            if (btn === '.') {
                if (currDisplay.includes('.') || currDisplay[-1] === '.') return;
                return dispatch(decimal(currDisplay, type));
            }
            return dispatch(lastKey_pressed(keyHistory, btn, type));
        }

        if (type === 'Command') {
            switch (btn) {
                case 'Clear':
                    return dispatch(clear_screen());
                case 'del':
                    return dispatch(del(currDisplay));
                case '=':
                    const newOperation = lastResult + currDisplay;
                    return dispatch(equals(newOperation));
                default:
                    return;
            }
        }

        if (type === 'Action' && lastType !== 'Action') {
            return dispatch(addOperator(currDisplay, btn, type));
        }
    }

    return (
        <button
            onClick={() => handleBtnClick(btn, type, lastType)}
            className={`btn ${operatorStyle ? 'btn--operator' : 'btn--number'}`}>
            {btn}
        </button >
    )
}

export default Button;
