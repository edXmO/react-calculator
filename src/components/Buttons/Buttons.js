import React from 'react';

// Components
import Button from './Button/Button';

// Styles 
import '../../styles/App.css';

const Buttons = ({ symbols }) => {
    const renderBtn = symbols.map(symbol => {
        const { key, type } = symbol;
        if (type === 'Operator') {
            return <Button key={key} btn={key} operatorStyle={'btn--operator'} />
        }
        if (type === 'Number') {
            return <Button key={key} btn={key} />
        }

    })

    return (
        <div className='buttons' >
            {renderBtn}
        </div>
    )
}

export default Buttons;


// Button component props
// onBtnClick={onBtnClick} action={action} 