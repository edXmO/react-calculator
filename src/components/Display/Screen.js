import React from 'react';


// Styles 
import '../../styles/App.css';

const Screen = ({ display, currentOperation }) => {
    return (
        <div className='screen'>
            <p className='screen__currOperation'>{currentOperation}</p>
            <p className='screen__display'>{display}</p>
        </div>
    )
}

export default Screen;