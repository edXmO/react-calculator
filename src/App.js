import React from 'react';
import './styles/App.css';

// Components
import Logo from './components/Logo/Logo';
import Screen from './components/Display/Screen';
import Buttons from './components/Buttons/Buttons';

const SYMBOLS = [
    { 'key': '+', 'type': 'Operator' },
    { 'key': '-', 'type': 'Operator' },
    { 'key': '/', 'type': 'Operator' },
    { 'key': 'Clear', 'type': 'Operator' },
    { 'key': '7', 'type': 'Number' },
    { 'key': '8', 'type': 'Number' },
    { 'key': '9', 'type': 'Number' },
    { 'key': 'x', 'type': 'Operator' },
    { 'key': '4', 'type': 'Number' },
    { 'key': '5', 'type': 'Number' },
    { 'key': '6', 'type': 'Number' },
    { 'key': 'e', 'type': 'Operator' },
    { 'key': '1', 'type': 'Number' },
    { 'key': '2', 'type': 'Number' },
    { 'key': '3', 'type': 'Number' },
    { 'key': '%', 'type': 'Operator' },
    { 'key': '0', 'type': 'Number' },
    { 'key': '.', 'type': 'Number' },
    { 'key': '<=', 'type': 'Number' },
    { 'key': '=', 'type': 'Number' }
];

const App = () => {

    return (
        <div className='main-wrapper'>
            <Logo />
            <Screen />
            <Buttons symbols={SYMBOLS} />
        </div>
    )
}

export default App;

// screen props
// display={screenValue} currentOperation={currentOperation}

// buttons props
// onBtnClick={getBtnClicked} 

 // const clearScreen = () => {
    //     setScreenValue('0');
    // }

    // const handleAction = action => {
    //     if (action === 'clear_screen') {
    //         clearScreen();
    //     }
    // }

    // const getBtnClicked = (btn, action) => {
    //     if (screenValue === '0') {
    //         setScreenValue(btn);
    //     }
    //     setScreenValue(prevValue => prevValue + btn);
    //     handleAction(action);
    // }

