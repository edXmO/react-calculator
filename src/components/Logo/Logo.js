import React from 'react';

// Styles 
import '../../styles/App.css';

const Logo = () => {
    return (
        <div className='logo' >
            Logo
        </div>
    )
}

export default Logo;