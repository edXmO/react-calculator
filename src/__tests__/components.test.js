import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

// Components
import Button from '../components/Buttons/Button/Button';
import Buttons from '../components/Buttons/Buttons';
import Copy from '../components/Copy/Copy';
import Screen from '../components/Display/Screen';
import Logo from '../components/Logo/Logo';
import App from '../App';
import SYMBOLS from '../models/symbolKeys';

let storeModel = {
    calculator: {
        result: '0',
        screen: '12+',
        lastType: 'Action',
        keyHistory: '12+',
        lastKey: '+'
    }
}

// Redux mock store
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);

// // Smoke bomb tests

// Tests 
// Button test
describe('Button component should render without crashing', () => {

    let store = mockStore(storeModel);

    store.dispatch = jest.fn();

    let component = renderer.create(
        <Provider store={store}>
            <Button type={'Display'} btn={'8'} operatorStyle={'btn--operator'} />
        </Provider>
    );

    it('should render with given state from redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
    it('should dispatch an action on button click', () => {
    })
});

// Tests
// Screen component
describe('Screen component should render without crashing', () => {
    let store = mockStore(storeModel);

    let component = renderer.create(
        <Provider store={store}>
            <Screen />
        </Provider>
    );
    it('should render with given state from redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should dispatch an action on button click', () => {

    })

});

describe('Buttons component should render without crashing', () => {
    let store = mockStore(storeModel);

    let component = renderer.create(
        <Provider store={store}>
            <Buttons symbols={SYMBOLS} />
        </Provider>
    );
    it('should render with given state from redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should dispatch an action on button click', () => {
    });
});


it('Copy component should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Copy />, div);
});

it('Logo component should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Logo />, div);
});


describe('App should render al components without crasing', () => {
    let store = mockStore(storeModel);

    let component = renderer.create(
        <Provider store={store}>
            <App />
        </Provider>
    );

    it('should render with given state from redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
})