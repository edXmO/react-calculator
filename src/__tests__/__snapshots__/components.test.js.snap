// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render al components without crasing should render with given state from redux store 1`] = `
<div
  className="main-wrapper"
>
  <div
    className="logo"
  >
    <h1
      className="logo__name"
    >
      CASIO
    </h1>
    <p
      className="logo__version"
    >
      fx-82MS
    </p>
  </div>
  <div
    className="screen"
  >
    <p
      className="screen__currOperation"
    >
      12+
    </p>
    <p
      className="screen__display"
    >
      0
    </p>
  </div>
  <div
    className="buttons"
  >
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      +
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      -
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      /
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      Clear
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      7
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      8
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      *
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      4
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      5
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      6
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      e
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      %
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      0
    </button>
    <button
      className="btn btn--number"
      onClick={[Function]}
    >
      .
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      del
    </button>
    <button
      className="btn btn--operator"
      onClick={[Function]}
    >
      =
    </button>
  </div>
</div>
`;

exports[`Button component should render without crashing should render with given state from redux store 1`] = `
<button
  className="btn btn--operator"
  onClick={[Function]}
>
  8
</button>
`;

exports[`Buttons component should render without crashing should render with given state from redux store 1`] = `
<div
  className="buttons"
>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    -
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    /
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    Clear
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    7
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    8
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    9
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    *
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    4
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    5
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    6
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    e
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    1
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    3
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    %
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    0
  </button>
  <button
    className="btn btn--number"
    onClick={[Function]}
  >
    .
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    del
  </button>
  <button
    className="btn btn--operator"
    onClick={[Function]}
  >
    =
  </button>
</div>
`;

exports[`Screen component should render without crashing should render with given state from redux store 1`] = `
<div
  className="screen"
>
  <p
    className="screen__currOperation"
  >
    12+
  </p>
  <p
    className="screen__display"
  >
    0
  </p>
</div>
`;
